### ISSUE TITLE
C++ and Py implementations of SpatialPooler differ in produced SDRs

### ISSUE THREAD

**Issue by breznak** [numenta/nupic-legacy#1352]
..when called directly. This has not been caught by `compatibility` tests. 
- the best, and most likely solution would be that I'm calling the code wrong. But I'm not aware of that. 
- I suspect `c++` version more, as Py learns faster and produces stable patterns for same input and different for different inputs.
- maybe just the defaults are not the same?

EDIT: I suspect there may be some SP instability problem (unrelated to cpp/py) where where multiple same runs result with significant difference! see 664c90c

TODO: revert 664c90c when fixed
TODO: enable https://github.com/numenta/nupic/pull/1351/files#diff-2de276b5266b55bfca508270faeb8ad3R392 when fixed - a testcase for this bug

CC @scottpurdy @oxtopus 

Blocked by: 
- [ ] #2290 same C++/Py params

**Comment by res1233** [comment]
Shouldn't the default config be stored centrally in a config file? That would solve this issue for good.

**Comment by rhyolight** [comment]
> Shouldn't the default config be stored centrally in a config file? That would solve this issue for good. Some JSON or XML config file should do the trick.

Yes, they are stored in https://github.com/numenta/nupic/blob/master/nupic/support/nupic-default.xml

**Comment by breznak** [comment]
@rhyolight I'm trying to unblock this with #2616 , unfortunately it does not seem to solve the problem. Could you review this issue and help to get some traction? I find it very odd to work with (c++) SP when we know the produced SDRs are not the same.

**Comment by rhyolight** [comment]
@subutai If, as @breznak says, the SDRs produced by our python implementation and our c++ implementation are truly producing different SDRs, would you consider that a serious problem that we need to investigate immediately? cc @numenta/nupic-committers

**Comment by cogmission** [comment]
Don't worry... (but investigate) it is probably due to the Random Number
Generation - Even though the basic algorithm (Mersenne Twister) is the
same, all the operations on the RNG don't yield the same results... Been
there, disgusted by that... :-)

On Mon, Oct 5, 2015 at 10:27 AM, Matthew Taylor <notifications@github.com>
wrote:

> @subutai <https://github.com/subutai> If, as @breznak
> <https://github.com/breznak> says, the SDRs produced by our python
> implementation and our c++ implementation are truly producing different
> SDRs, would you consider that a serious problem that we need to investigate
> immediately? cc @numenta/nupic-committers
> <https://github.com/orgs/numenta/teams/nupic-committers>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numenta/nupic/issues/1352#issuecomment-145569649>.
>



-- 
*With kind regards,*

David Ray
Java Solutions Architect

*Cortical.io <http://cortical.io/>*
Sponsor of:  HTM.java <https://github.com/numenta/htm.java>

d.ray@cortical.io
http://cortical.io

**Comment by subutai** [comment]
> @subutai If, as @breznak says, the SDRs produced by our python implementation and our c++ implementation are truly producing different SDRs, would you consider that a serious problem that we need to investigate immediately? cc @numenta/nupic-committers

Yes, this would be good to investigate. We do have extensive tests on this and in those tests the two implementations are identical. I do know that they can start to depart for long running jobs due to precision differences in Python/C++ math libraries.  

I can't judge the priority unless we know the exact situation. Is there a simple test script that can reproduce the differences? If I understand the comments above it could simply be that the default constructors are different and should be the same. (See my Sep 23 2014 comment.) Is there any other information? 

BTW, it can't be due to the RNG code because both implementations call the exact same piece of code. The difference has to be outside of the RNG implementation.

**Comment by rcrowder** [comment]
@chetan51 Have your recent changes corrected this issue, or is this a WIP super-issue and reliant on #2290? cc @breznak

**Comment by chetan51** [comment]
My recent changes should have corrected this issue. @breznak can you confirm?

**Comment by breznak** [comment]
@chetan51 thanks! 
I cannot confirm now, will have time later next week. But from the past the remaining items here were: 

> TODO: revert 664c90c when fixed
TODO: enable https://github.com/numenta/nupic/pull/1351/files#diff-2de276b5266b55bfca508270faeb8ad3R392 when fixed - a testcase for this bug
CC @scottpurdy @oxtopus
Blocked by:
    #2290 same C++/Py params

I'll need to test that and understand the motivation for this issue again a bit. 
PS: I think #2290 will still be a problem.
